### ISSUE TITLE
VariableSubstitutionPreprocessor is not working

### ISSUE THREAD

**Issue by vlavellauruit** [DbUp/DbUp#167]
Hi everybody, 

I'm trying to trigger a variable preprocessor, but I get this error on my  excecution: 

"Variable XX has not value defined"

I wrote this code:

```
var upgrader =
DeployChanges.To.SqlDatabase(connectionString)
                .WithPreprocessor(new VariableSubstitutionPreprocessor(
                new Dictionary<string, string>() { { "'candy'", "'m&m'" } }))                        .WithScriptsAndCodeEmbeddedInAssembly(Assembly.GetExecutingAssembly())
                .WithTransaction()          
                .LogToConsole()
                .Build();
```

Should I change something, or this should work? I didn't find something on the docs that indicates to me that I'm doing something worng

Thanks

**Comment by omnibs** [comment]
**TLDR; don't use singlequotes in variable names; use `WithVariables()` instead of `WithPreprocessor()`.**

I thought this might be a problem with the single quotes so I tried `WithPreprocessor(new VariableSubstitutionPreprocessor())` with simple variables. It didn't work. Seems to be a bug in this form. haven't seen it in the docs though, so IDK. @JakeGinnivan? Is this a "won't fix"?

So I tried the recommended syntax from the docs, and it didn't work with single quotes. I got `SqlException (0x80131904): Incorrect syntax near 'candy'`.

The recommended syntax works as expected with unquoted variables tho:

```sql
create table $tableName$ (a int not null);
```

```c#
var upgrader =
                DeployChanges.To
                    .SqlDatabase(connectionString)
                    .WithScriptsEmbeddedInAssembly(Assembly.GetExecutingAssembly())
                    .WithVariable("tableName", "bergamot")
                    .LogToConsole()
                    .Build();
```

Also, for Dictionaries, like in your question:

```
var upgrader =
                DeployChanges.To
                    .SqlDatabase(connectionString)
                    .WithScriptsEmbeddedInAssembly(Assembly.GetExecutingAssembly())
                    .WithVariables(new Dictionary<string, string>
                    {
                        {"tableName", "bergamot" }
                    })
                    .LogToConsole()
                    .Build();
```

I'll leave this up for @JakeGinnivan to give us his 5cent whether the WithPreprocessor is a bug or a wontfix.

**Comment by JakeGinnivan** [comment]
Maybe we should just strip single quotes out automatically, or if they are present give a warning. Or add xml docs to say don't use quotes.

There are plenty of things we could do to improve this. @vlavellauruit which of the above suggestions (or others) would work for you? Happy to receive a PR from anyone making this better
